### PR TITLE
Core/AdventureJournal: Add 3 Handle OpenQuest, StartQuest and UpdateSuggestions with AdventureJournal.db2

### PR DESCRIPTION
**Pull request by osleyder85**
**Changes proposed:**

-  Add HandleAdventureJournalOpenQuest
-  Add HandleAdventureJournalStartQuest
-  Add HandleAdventureJournalUpdateSuggestions
-  Add WorldPacket AdventureJournalOpenQuest
-  Add WorldPacket AdventureJournalStartQuest
-  Add WorldPacket AdventureJournalUpdateSuggestions
-  Add WorldPacket AdventureJournalDataResponse
-  Add AdventureJournal.db2

**Target branch(es):** 3.3.5/master

- [ ] 3.3.5
- [X] master

**Issues addressed:**

Closes #  (insert issue tracker number)

**Tests performed:**

(Does it build, tested in-game, etc.)

**Known issues and TODO list:** (add/remove lines as needed)

- [X] it build
- [X] tested in-game
mo

### CODE REVIEW

**Comment by Faq** [comment]
Just info: As this is work in progress, for such cases github has draft status for PRs.

File: src/server/game/Server/Packets/MiscPackets.cpp [hunk 1]
```diff
@@ -711,3 +711,34 @@ WorldPacket const* WorldPackets::Misc::StartTimer::Write()
...
+}
+
+void WorldPackets::Misc::AdventureJournalUpdateSuggestions::Read()
+{
+    OnLevelUp = _worldPacket.ReadBit();
+}
+
+WorldPacket const* WorldPackets::Misc::AdventureJournalDataResponse::Write()
+{
+    _worldPacket.WriteBit(OnLevelUp1);
```
Comment (MaxtorCoder) [inline src/server/game/Server/Packets/MiscPackets.cpp:732]:
```suggestion
    _worldPacket.WriteBit(OnLevelUp);
```

File: src/server/game/Server/Packets/MiscPackets.cpp [hunk 2]
```diff
@@ -711,3 +711,34 @@ WorldPacket const* WorldPackets::Misc::StartTimer::Write()
...
+    _worldPacket.FlushBits();
+    _worldPacket << NumEntries;
+    _worldPacket << uint32(AdventureJournalDatas.size());
```
Comment (MaxtorCoder) [inline src/server/game/Server/Packets/MiscPackets.cpp:735]:
```suggestion
    _worldPacket << uint32(AdventureJournalDatas.size());
```

File: src/server/game/Handlers/MiscHandler.cpp [hunk 3]
```diff
@@ -1144,3 +1144,22 @@ void WorldSession::HandleCloseInteraction(WorldPackets::Misc::CloseInteraction&
...
+void WorldSession::HandleAdventureJournalOpenQuest(WorldPackets::Misc::AdventureJournalOpenQuest& packet)
+{
+    if (AdventureJournalEntry const* adventureJournalEntry = sAdventureJournalStore.LookupEntry(packet.AdventureJournalID))
+        if (Quest const* quest = sObjectMgr->GetQuestTemplate(adventureJournalEntry->QuestID))
+            if (!_player->hasQuest(adventureJournalEntry->QuestID) && _player->CanTakeQuest(quest, true))
+                if (WorldSession* session = _player->GetSession())
+                {
+                    PlayerMenu menu(session);
+                    menu.SendQuestGiverQuestDetails(quest, _player->GetGUID(), true, false);
+                }
```
Comment (MaxtorCoder) [inline src/server/game/Handlers/MiscHandler.cpp:1157]:
```suggestion
    auto adventureJournalEntry = sAdventureJournalStore[packet.AdventureJournalID];
    if (!adventureJournalEntry)
        return;

    Quest const* quest = sObjectMgr->GetQuestTemplate(adventureJournalEntry->QuestID);
    if (!quest)
        return;

    if (_player->HasQuest(adventureJournalEntry->QuestID) && _player->CanTakeQuest(quest, true))
    {
        PlayerMenu menu(_player->GetSession());
        menu.SendQuestGiverQuestDetails(quest, _player->GetGUID(), true, false);
    }
```

File: src/server/game/Server/Packets/MiscPackets.h [hunk 4]
```diff
@@ -919,6 +919,54 @@ namespace WorldPackets
...
+
+            WorldPacket const* Write() override;
+
+            struct AdventureJournalDataInfo
+            {
+                int32 AdventureJournalID = 0;
+                int32 Priority = 0;
+            };
+
+            bool OnLevelUp1 = false;
```
Comment (MaxtorCoder) [inline src/server/game/Server/Packets/MiscPackets.h:966]:
```suggestion
            bool OnLevelUp = false;
```

File: src/server/game/Server/Packets/MiscPackets.h [hunk 5]
```diff
@@ -919,6 +919,54 @@ namespace WorldPackets
...
+            int32 NumEntries = 0;
+            std::vector<AdventureJournalDataInfo> AdventureJournalDatas;
```
Comment (MaxtorCoder) [inline src/server/game/Server/Packets/MiscPackets.h:968]:
```suggestion
            std::vector<AdventureJournalDataInfo> AdventureJournalDatas;
```

**Review by MaxtorCoder — CHANGES_REQUESTED** [review]

File: src/server/game/Handlers/MiscHandler.cpp [hunk 6]
```diff
@@ -1144,3 +1144,27 @@ void WorldSession::HandleCloseInteraction(WorldPackets::Misc::CloseInteraction&
...
+        PlayerMenu menu(_player->GetSession());
+        menu.SendQuestGiverQuestDetails(quest, _player->GetGUID(), true, false);
+    }
+}
+
+void WorldSession::HandleAdventureJournalStartQuest(WorldPackets::Misc::AdventureJournalStartQuest& packet)
+{
+    if (Quest const* quest = sObjectMgr->GetQuestTemplate(packet.QuestID))
+        if (!_player->hasQuest(packet.QuestID))
+            _player->AddQuestAndCheckCompletion(quest, nullptr);
```
Comment (mdX7) [inline src/server/game/Handlers/MiscHandler.cpp:1169]: possible exploit - you gotta check whether the quest is valid for adventureguide or not (possibly some flag), would also check that in HandleAdventureJournalOpenQuest.

NEVER trust client.

File: src/server/game/Server/Packets/MiscPackets.cpp [hunk 7]
```diff
@@ -711,3 +711,32 @@ WorldPacket const* WorldPackets::Misc::StartTimer::Write()
 
     return &_worldPacket;
 }
+
+void WorldPackets::Misc::AdventureJournalOpenQuest::Read()
+{
+    _worldPacket >> AdventureJournalID;
```
Comment (mdX7) [inline src/server/game/Server/Packets/MiscPackets.cpp:717]: uses casts like you did with .size()

e.g.
```c
_worldPacket << uint32(AdventureJournalID);
```
in this case

do this for all primitive datatypes

File: src/server/game/Server/Packets/MiscPackets.cpp [hunk 8]
```diff
@@ -711,3 +711,32 @@ WorldPacket const* WorldPackets::Misc::StartTimer::Write()
...
+}
+
+WorldPacket const* WorldPackets::Misc::AdventureJournalDataResponse::Write()
+{
+    _worldPacket.WriteBit(OnLevelUp);
+    _worldPacket.FlushBits();
+    _worldPacket << uint32(AdventureJournalDatas.size());
+    for (auto const& dataInfo : AdventureJournalDatas)
+    {
+        _worldPacket << dataInfo.AdventureJournalID;
```
Comment (mdX7) [inline src/server/game/Server/Packets/MiscPackets.cpp:737]: uses casts like you did with .size()

e.g.
```c
_worldPacket << uint32(dataInfo.AdventureJournalID);
```
in this case

do this for all primitive datatypes

File: src/server/game/Handlers/MiscHandler.cpp [hunk 9]
```diff
@@ -1144,3 +1144,27 @@ void WorldSession::HandleCloseInteraction(WorldPackets::Misc::CloseInteraction&
...
+{
+    auto adventureJournalEntry = sAdventureJournalStore[packet.AdventureJournalID];
+    if (!adventureJournalEntry)
+        return;
+
+    Quest const* quest = sObjectMgr->GetQuestTemplate(adventureJournalEntry->QuestID);
+    if (!quest)
+        return;
+
+    if (_player->HasQuest(adventureJournalEntry->QuestID) && _player->CanTakeQuest(quest, true))
```
Comment (MaxtorCoder) [inline src/server/game/Handlers/MiscHandler.cpp:1158]:
```suggestion
    if (_player->hasQuest(adventureJournalEntry->QuestID) && _player->CanTakeQuest(quest, true))
```
My dumbass accidentally capitalized this

**Comment by Shauren** [comment]
I'm not sure why you unmarked this as draft, you did not adress the exploit comment AT ALL

**Comment by osleyder85** [comment]
The error was caused by applying a suggestion made by MaxtorCode. I already corrected the error.

**Comment by mdX7** [comment]
> 
> 
> The error was caused by applying a suggestion made by MaxtorCode. I already corrected the error.

you missed my comments

**Comment by osleyder85** [comment]
> 
> 
> > The error was caused by applying a suggestion made by MaxtorCode. I already corrected the error.
> 
> you missed my comments

Oh, sorry, now I get to work to fix what I recommend. I'm still sleeping.

**Comment by MaxtorCoder** [comment]
The latest commit does not fix any exploit..?

File: src/server/game/Server/Packets/MiscPackets.cpp [hunk 10]
```diff
@@ -711,3 +711,32 @@ WorldPacket const* WorldPackets::Misc::StartTimer::Write()
...
+}
+
+WorldPacket const* WorldPackets::Misc::AdventureJournalDataResponse::Write()
+{
+    _worldPacket.WriteBit(OnLevelUp);
+    _worldPacket.FlushBits();
+    _worldPacket << uint32(AdventureJournalDatas.size());
+    for (auto const& dataInfo : AdventureJournalDatas)
+    {
+        _worldPacket << uint32(dataInfo.AdventureJournalID);
```
Comment (mdX7) [inline src/server/game/Server/Packets/MiscPackets.cpp:737]: int32, as defined in WorldPackets::Misc::AdventureJournalDataResponse

File: src/server/game/Server/Packets/MiscPackets.cpp [hunk 11]
```diff
@@ -711,3 +711,32 @@ WorldPacket const* WorldPackets::Misc::StartTimer::Write()
...
+        _worldPacket << dataInfo.Priority;
```
Comment (mdX7) [inline src/server/game/Server/Packets/MiscPackets.cpp:738]: int32 cast as stated in WorldPackets::Misc::AdventureJournalDataResponse

File: src/server/game/Handlers/MiscHandler.cpp [hunk 12]
```diff
@@ -1144,3 +1144,30 @@ void WorldSession::HandleCloseInteraction(WorldPackets::Misc::CloseInteraction&
...
+}
+
+void WorldSession::HandleAdventureJournalStartQuest(WorldPackets::Misc::AdventureJournalStartQuest& packet)
+{
+    Quest const* quest = sObjectMgr->GetQuestTemplate(packet.QuestID)
+    if (!quest)
+        return;
+
+    if (!_player->hasQuest(packet.QuestID))
+        _player->AddQuestAndCheckCompletion(quest, nullptr);
```
Comment (mdX7) [inline src/server/game/Handlers/MiscHandler.cpp:1172]: this did not fix any exploit. You either have to check if the questid is in adventurejournal db2 or maybe theres even a fitting flag for it

**Review by mdX7 — CHANGES_REQUESTED** [review]

File: src/server/game/Server/Packets/MiscPackets.cpp (hunk 10 above)
Comment (osleyder85) [reply src/server/game/Server/Packets/MiscPackets.cpp:737]: Now it was that I understood what they were saying to me. Yes, of course I cast it with the same type of data.

File: src/server/game/Handlers/MiscHandler.cpp [hunk 13]
```diff
@@ -1144,3 +1144,34 @@ void WorldSession::HandleCloseInteraction(WorldPackets::Misc::CloseInteraction&
...
+    }
+}
+
+void WorldSession::HandleAdventureJournalStartQuest(WorldPackets::Misc::AdventureJournalStartQuest& packet)
+{
+    Quest const* quest = sObjectMgr->GetQuestTemplate(packet.QuestID);
+    if (!quest)
+        return;
+
+    auto adventureJournalEntry = sAdventureJournalStore[quest->GetQuestId()];
```
Comment (MaxtorCoder) [inline src/server/game/Handlers/MiscHandler.cpp:1171]: Not how it works. Loop over `sAdventureJournal` and check QuestID then.

**Review by MaxtorCoder — CHANGES_REQUESTED** [review]

**Comment by mdX7** [comment]
now the last thing you gotta check is playercondition which is referenced in AdventureGuide.db2 for both opcodes after getting AdventureGuide entry and then we're good i think. (otherwise we might get M+ dungeon quest when it's not even up, we dont want that right?
Ddoes playercondition include specific world events e.g. timewalking? Otherwise that needs handling aswell.

**Comment by Shauren** [comment]
I still see this as incomplete, you added packet structures for `CMSG_ADVENTURE_JOURNAL_UPDATE_SUGGESTIONS`, `SMSG_ADVENTURE_JOURNAL_DATA_RESPONSE` and enabled them but added no handlers

**Comment by osleyder85** [comment]
When the player levels up and interacts with the Adventure Guide menu, the server crashes. I can't find the problem
causing the server to crash. [crashes_worldserver.txt](https://pastebin.com/9xe0N7jA)

**Comment by MaxtorCoder** [comment]
> When the player levels up and interacts with the Adventure Guide menu, the server crashes. I can't find the problem
> causing the server to crash. [crashes_worldserver.txt](https://pastebin.com/9xe0N7jA)

Stol building in release and build in RelWithDebInfo. Logs are useless in release.

**Comment by Shauren** [comment]
`hasQuest` is not a correct function to use. It checks if the object is a questgiver that gives this quest, not if player has this quest in his log

**Comment by osleyder85** [comment]
Fixed server crash.

**Comment by osleyder85** [comment]
@Shauren , you think something else is missing from PR.

**Comment by osleyder85** [comment]
For my part, everything is ready.

**Comment by matanshukry** [comment]
Great PR!

**Comment by Shauren** [comment]
Merged without CMSG_ADVENTURE_JOURNAL_START_QUEST in 3addbe7dd418259149c39fc3b9423f7809de696a - it validates quest using incorrect db2 file

**Pull request closed without merging**